### DISCUSSION TITLE
Mitmproxy and me being a newbie + client cert auth + access to CA

### DISCUSSION THREAD

**Discussion by mihaime** [mitmproxy/mitmproxy#5848] [Q&A]
Hi,

I've recently installed mitmproxy in my setup and I'm having a bit of a hard time wrapping my ears around it it seems.
I have to mention that I do control all elements in the path so if I really wanted to, I could intercept the traffic.

Setup is:
- Client ---> sshl (multiplexer) --> Server
- sslh is exposing 3 services (2 x plain HTTPs based, 1 is Grpc -> http/2)
- client and server do mutual auth one to the other
- server sends its own Cert (signed by Root CA)
- client sends its own Cert + Bundle (cert is signed by intermediate CA, intermediate is signed by Root CA, same as above)

I tried to install mitmproxy on the server and do the following:
`mitmproxy --certs client-hostname=/path/to/client_cert/client.pem --certs server-hostname=/path/to/server_cert/server.pem' --set block_global=false --mode transparent --set add_upstream_certs_to_client_chain=true`

I made sure that client.pem has:
- private key + then client certificate
- intermediate CA + then root ca

I looked that server.pem has:
- private key + server certificate
- root ca (this one was not signed by the intermediate)

What I see is:
`asn1 encoding routines, wrong tag`
`asn1 encodinging routines, nested asn1 error`
and that the default CA of mitmproxy gets used to generate the "server" certificate (which I assume it is so because it cannot load the certs I specified).

Any hint here or is there also some other param to mitmproxy that I have to look at for this mutual client cert auth + not using the default CA?
(I did not notice that if I replace the ~/.mitmproxy/mitmproxy-ca.pem with my own then I don't get my "server" cert used but at least the generated one is using the trusted CA)

Sorry for being so rookie with this.
Hope I can get 1-2 points as I'm banging my head against the wall and feel a bit hopeless.

**Comment by mhils** [comment]
For client certificates, make sure to specify `client_certs` and not `certs`: https://docs.mitmproxy.org/dev/concepts-certificates/#using-a-client-side-certificate

**Reply by mihaime** [reply]
@mhils  sorry to be stressing.
If I get it right:
- for the client cert I use "client_certs" 
- for the server cert side I would use "certs" so that I do not end up with one generated either by the "embedded" root CA or by mine (if I replace that file of the custom ca) ?

Asking as while that fixed the logic, I still get the:
`asn1 encoding routines, wrong tag
asn1 encodinging routines, nested asn1 error`

**Reply by mhils** [reply]
Yes, correct. I unfortunately have no idea about your asn1 error. This looks like a malformed certificate?

**Reply by mihaime** [reply]
What I also assumed but when I tried to print everything in both certs (client/server) using this:
` openssl crl2pkcs7 -nocrl -certfile ~/client_certs/<client|server>.pem | openssl pkcs7 -print_certs -text -noout`
everything displayed just fine.

That's where I got stuck in fact as I did not get from the ssl lib part any clear error as to what it does not like.
I even tried moving things up / down (chain wise) in the pem file thinking maybe it's some parsing thing.
Any constrains as to private key format or something else?
Mine is RSA in there.

**Comment by jay00089** [comment]
Is there any way to do tls mutual authentication with mitm proxy?

**Reply by mhils** [reply]
Yes, see the option mentioned above.

**Comment by jay00089** [comment]
I have tried to make mutual authentication using the above code but i'm getting the below error  
Client TLS handshake failed. The client may not trust the proxy's certificate for <no address> (OpenSSL Error([('SSL routines', '', 'inappropriate fallback')]))

can u help me resolve it

**Comment by jay00089** [comment]
Does MiTM proxy support mutual authentication between Proxy and Client? All i need is to get the client certificate by sending a "CERTIFICATE_REQUEST" TLS handshake message from proxy. Is it possible achieve by any available options.
